README.md
![ci-cd](https://github.com/UBC-MDS/colourpycker/actions/workflows/ci-cd.yml/badge.svg)
[![codecov](https://codecov.io/gh/UBC-MDS/colourpycker/branch/main/graph/badge.svg?token=pkGKQ2l0uG)](https://codecov.io/gh/UBC-MDS/colourpycker)
[![Documentation Status](https://readthedocs.org/projects/colourpycker/badge/?version=latest)](https://colourpycker.readthedocs.io/en/latest/?badge=latest)

# colourpycker

`colourpycker` is a Python package that can be used to extract colours from images for use in data visualization projects.

## Overview

This package allows users to integrate unique colour palettes into their graphs for exploratory data analysis. The colours are retrieved from image data (via URL) and are selected based on their overall prominence in a picture. While there are existing tools that are used to process images and create figures independently, we aim to combine both of their functionalities to help programmers easily design effective and creative visualizations.

## Where Our Package Fits

As mentioned, there are packages in Python that are capable of colour extraction and data visualization, but none exist that combine the two functions to our knowledge. Some examples of such packages are as follows:

[Pillow](https://pypi.org/project/Pillow/): This package adds image processing capabilities into Python interpreters. It can perform various image transformations but does not allow for colours to be extracted directly for further use.

[extcolors](https://pypi.org/project/extcolors/): This extracts RGB colour codes from images into text along with the occurrence rate (proportion of pixels). However, we would need to use additional packages to create plots using common colours in the image. This would also not allow us to invert them, or create a palette.

[color-extraction](https://pypi.org/project/color-extraction/): This provides the most similar color for the pixels of an image from a palette of predefined colors.

[Pylette](https://github.com/qTipTip/Pylette/): This extracts colours and creates a colour palette based on an input image. This palette can be dumped into a csv file that would then need to be processed for use in data visualizations.

Therefore, we are developing `colourpycker` to bridge this gap.

## Installation

```bash
$ pip install colourpycker
```

## Functions

`get_color_palette(img_url, tolerance, limit)`: This function extracts the most common colors from an image and returns them as a data frame of hex color codes and RGB values. The user is provided with the ability to set tolerance while picking colors, along with the number of colors that should be returned.

`donut(img_url, num_clrs, tolerance, img_size, plot_show=True)`: This function creates a donut chart with information on the $n$ most common colors in the linked image. The user specifies the image, the number of colors, the tolerance for how similar the colors will be, and specifies the size of the resulting chart in pixels.

`scatterplot(img_url, dataset, x, y, fill, tolerance)`: This function creates a simple scatterplot using the colours selected from the image based on a dataset of the users choosing.

`negative(img_url, num_colours, tolerance)`: This function extracts the most common $n$ colours from an image (via URL) and inverts them to retrieve the negative version of a colour palette, returning the associated HEX codes and RGB values for each colour.

## Usage

For more examples on how to use `colourpycker`, a Jupyter Notebook can be found [here](https://github.com/UBC-MDS/colourpycker/blob/main/docs/example.ipynb) with additional use cases.

```
from colourpycker.colourpycker import get_colour_palette, donut, scatterplot, scatterplot, negative, get_color_palette, donut

get_color_palette('https://i.imgur.com/s9egWBB.jpg', 20, 5)

donut('https://i.imgur.com/s9egWBB.jpg', 5, 20, 400)

scatterplot('https://i.imgur.com/s9egWBB.jpg', penguins, 'bill_length_mm', 'body_mass_g', 'species', 50)

negative("https://i.imgur.com/s9egWBB.jpg", 10, 20)
```

## Contributing

Interested in contributing? Check out the [contributing guidelines](https://github.com/UBC-MDS/colourpycker/blob/main/CONTRIBUTING.md) and [list of contributors](https://github.com/UBC-MDS/colourpycker/blob/main/CONTRIBUTORS.md) who worked on this project. Please note that this project is released with a [Code of Conduct](https://github.com/UBC-MDS/colourpycker/blob/main/CONDUCT.md). By contributing to this project, you agree to abide by its terms.

## License

`colourpycker` was created by Shaun Hutchinson, Arjun Radhakrishnan, Alex Taciuk, and Lauren Zung. It is licensed under the terms of the MIT license.

## Documentation

The official documentation is hosted on Read the Docs: <https://colourpycker.readthedocs.io/en/latest/example.html>

## Credits

`colourpycker` was created with [`cookiecutter`](https://cookiecutter.readthedocs.io/en/latest/) and the `py-pkgs-cookiecutter` [template](https://github.com/py-pkgs/py-pkgs-cookiecutter).


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.0.6 (2023-02-04)


## v0.0.5 (2023-01-29)


## v0.0.4 (2023-01-29)


## v0.0.3 (2023-01-29)


## v0.0.2 (2023-01-27)


## v0.0.1 (2023-01-26)


## v0.1.0 (10/01/2023)

- First release of `colourpycker`!

tests/test_colourpycker.py
import numpy as np
import pandas as pd
import altair as alt
import pandas as pd
from colourpycker.colourpycker import donut, get_color_palette, scatterplot, negative
import matplotlib.text as text
import pytest


def test_get_color_palette_valid_input():
    """Test for get_color_palette when input parameters are valid"""
    img_url = "https://i.imgur.com/s9egWBB.jpg"
    tolerance = 20
    limit = 5
    result = get_color_palette(img_url, tolerance, limit)
    assert isinstance(result, pd.DataFrame)
    assert result.shape == (5, 3)


def test_get_color_palette_not_a_url():
    """Test for get_color_palette when URL is a random string"""
    try:
        img_url = "soijfoidjfosdijfoisdj"
        tolerance = 20
        limit = 5
        result = get_color_palette(img_url, tolerance, limit)
        assert False
    except:
        # Make sure no exception is raised and the function handles it
        assert True


def test_get_color_palette_url_but_not_image():
    """Test for get_color_palette when URL is valid but it is not pointing to an image"""
    try:
        img_url = "https://soijfoidjfosdijfoisdj.url"
        tolerance = 20
        limit = 5
        result = get_color_palette(img_url, tolerance, limit)
        assert False
    except:
        # Make sure no exception is raised and the function handles it
        assert True


def test_get_color_palette_invalid_tolerance():
    """Test for get_color_palette when tolerance is invalid"""
    try:
        img_url = "https://i.imgur.com/s9egWBB.jpg"
        tolerance = -20000
        limit = 5
        result = get_color_palette(img_url, tolerance, limit)
    except:
        # Make sure no exception is raised and the function handles it
        assert True


def test_donut():
    """Testing that donut() generates the correct output"""
    actual = str(donut("https://i.imgur.com/s9egWBB.jpg", 5, 30, 200, plot_show=False).findobj(text.Text)[0:5]).count("#")
    expect = 5

    assert actual == expect, "the function is not plotting the correct amount of colors"

    actual = str(
        type(
            donut(
                "https://i.imgur.com/s9egWBB.jpg",
                5,
                30,
                200,
                plot_show=False,
            )
        )
    )
    expect = "<class 'matplotlib.figure.Figure'>"
    assert actual == expect, "The function is not outputting a matplotlib figure"

        
    actual = str(donut("https://i.imgur.com/s9egWBB.jpg", 5, 30, 200, plot_show=False).findobj(text.Text)[0]).split("#")[1].split(":")[0]
    expect = get_color_palette("https://i.imgur.com/s9egWBB.jpg", 30, 5).iloc[0,0].replace("#","")
    assert actual == expect, "the function is not returning the correct colors"

def test_scatterplot():
    """Testing that the function scatterplot() generates the correct output and expects the right inputs."""
    # Creating a test dataframe
    test_df = pd.DataFrame(
        {
            "x": np.array([1.1, 1.3, 2.4, 1.5, 1.9, 2.1, 2.1, 0.9, 3.1]),
            "y": np.array([6.8, 7.2, 9.4, 8.3, 5.4, 9.6, 2.9, 6.6, 4.3]),
            "fill": np.array(["A", "B", "C", "A", "B", "C", "A", "B", "C"]),
        }
    )

    # Creating a test plot
    test_scatter = scatterplot(
        "https://visit.ubc.ca/wp-content/uploads/2019/04/plantrip_header-2800x1000_2x.jpg",
        test_df,
        "x",
        "y",
        "fill",
        50,
    )

    # Tests
    assert test_scatter.mark == "point", "The mark should be a point"

    assert (
        type(test_scatter) == alt.vegalite.v4.api.Chart
    ), "The return should be an altair Chart"

    assert (
        test_scatter.encoding.fill.scale.range
        == get_color_palette(
            r"https://visit.ubc.ca/wp-content/uploads/2019/04/plantrip_header-2800x1000_2x.jpg",
            50,
            3,
        )["HEX"].to_list()
    ), "The fill of the points should be the colours from the image"

    assert (
        test_scatter.encoding.x.field == "x"
    ), "The x-axis of the points should be mapped to the x axis of the Chart"

    assert (
        test_scatter.encoding.y.field == "y"
    ), "The y-axis of the points should be mapped to the y axis of the Chart"

    with pytest.raises(TypeError):
        # x cannot be numeric
        scatterplot(
            "https://visit.ubc.ca/wp-content/uploads/2019/04/plantrip_header-2800x1000_2x.jpg",
            test_df,
            50,
            "y",
            "fill",
            50,
        )
        # y cannot be numeric
        scatterplot(
            "https://visit.ubc.ca/wp-content/uploads/2019/04/plantrip_header-2800x1000_2x.jpg",
            test_df,
            "x",
            50,
            "fill",
            50,
        )
        # fill cannot be numeric
        scatterplot(
            "https://visit.ubc.ca/wp-content/uploads/2019/04/plantrip_header-2800x1000_2x.jpg",
            test_df,
            "x",
            "y",
            "50",
            50,
        )
        # tolerance cannot be cannot a float
        scatterplot(
            "https://visit.ubc.ca/wp-content/uploads/2019/04/plantrip_header-2800x1000_2x.jpg",
            test_df,
            "x",
            "y",
            "50",
            50.75,
        )

def test_negative():
    """Test that negative() generates the correct output and expects the right inputs."""
    expected = pd.DataFrame(
        {
            "HEX": ["#94604a", "#f6edf2", "#244176", "#c999d8", "#3eaadc"],
            "RGB": [
                (148, 96, 74),
                (246, 237, 242),
                (36, 65, 118),
                (201, 153, 216),
                (62, 170, 220),
            ],
        }
    )

    actual = negative(
        "https://i.imgur.com/s9egWBB.jpg",
        5,
        50,
    )

    assert isinstance(actual, pd.DataFrame), "Output is not a data frame."
    assert (
        actual.columns[0] == expected.columns[0]
    ), "Output data frame does not contain the column for HEX codes."
    assert (
        len(actual.index) == 5
    ), "Output data frame does not contain the correct number of rows."
    assert (
        expected["HEX"].iloc[0] == actual["HEX"].iloc[0]
    ), "HEX code of the most common negative colour is not correct."
    assert (
        expected["RGB"].iloc[0] == actual["RGB"].iloc[0]
    ), "RGB code of the most common negative colour is not correct."

    # check inversion
    test_rgb = expected["RGB"].iloc[2]
    normal_rgb = (
        255 - int(test_rgb[0]),
        255 - int(test_rgb[1]),
        255 - int(test_rgb[2]),
    )

    actual_rgb = actual["RGB"].iloc[2]
    actual_normal = (
        255 - int(actual_rgb[0]),
        255 - int(actual_rgb[1]),
        255 - int(actual_rgb[2]),
    )

    assert actual_normal == normal_rgb, "RGB code did not invert correctly."

    # maximum tolerance
    high_tol = negative(
        "https://i.imgur.com/s9egWBB.jpg",
        10,
        100,
    )
    # tolerance was set to a high value relative to the number of colours extracted
    moderate_tol = negative(
        "https://i.imgur.com/s9egWBB.jpg",
        20,
        88,
    )

    assert (
        len(high_tol.index) == 1
    ), "Setting tolerance to 100 should result in a maximum of 1 colour extracted."
    assert (
        len(moderate_tol.index) <= 20
    ), "Output data frame should have less than the specified number of colours."

    # partially transparent image
    partial = negative("https://i.stack.imgur.com/G4NCG.png", 20, 5)

    assert (
        len(partial.index) == 1
    ), "Output data frame should contain one colour; partially transparent images should ignore transparent pixels."

    with pytest.raises(ValueError):
        # trying to read a path instead of a URL
        negative("path/to/img.png", 4, 10)
        # trying to extract colours from a transparent image
        negative(
            "https://upload.wikimedia.org/wikipedia/commons/8/89/HD_transparent_picture.png",
            5,
            100,
        )

    with pytest.raises(TypeError):
        # tolerance cannot be a float
        negative(
            "https://i.imgur.com/s9egWBB.jpg",
            4,
            25.5,
        )
        # number of colours cannot be a float
        negative(
            "https://i.imgur.com/s9egWBB.jpg",
            4.8,
            25,
        )
        # tolerance must be less than 100
        negative(
            "https://i.imgur.com/s9egWBB.jpg",
            5,
            101,
        )
        # tolerance cannot be 0
        negative(
            "https://i.imgur.com/s9egWBB.jpg",
            5,
            0,
        )
        # tolerance cannot be a negative integer
        negative(
            "https://i.imgur.com/s9egWBB.jpg",
            5,
            -10,
        )

colourpycker/tests/test_colourpycker.py
from colourpycker import colourpycker


src/colourpycker/colourpycker.py
import pandas as pd
import requests
from io import BytesIO
from extcolors import extract_from_path
import PIL
from PIL import Image
import matplotlib.pyplot as plt
import extcolors
import altair as alt
import os
import re
import requests


def check_param_validity(url, tolerance, limit):
    """
    Checks if the input params are valid

    Parameters
    ----------
    img_url: str
        the url of the image from which the color palette is to be extracted
    tolerance: int
        a value between 0 and 100 representing the tolerance level for color matching
    limit: int
        the maximum number of colors to be returned in the palettel

    Returns:
    -------
    bool:
        True if params are valid else return False
    """

    # Regex from https://www.geeksforgeeks.org/check-if-an-url-is-valid-or-not-using-regular-expression/

    regex = r"(?i)\b((?:https?://|www\d{0,3}[.]|[a-z0-9.\-]+[.][a-z]{2,4}/)(?:[^\s()<>]+|\(([^\s()<>]+|(\([^\s()<>]+\)))*\))+(?:\(([^\s()<>]+|(\([^\s()<>]+\)))*\)|[^\s`!()\[\]{};:'\".,<>?«»“”‘’]))"
    return (
        re.search(regex, url) is not None
        and (tolerance >= 0 and tolerance <= 100)
        and (limit >= 0 and limit <= 100)
    )


def rgb_to_hex(rgb):
    """
    Convert an RGB value stored as a tuple to a hex color code.

    Parameters
    ----------
    rgb : tuple
        The tuple of integers representing the RGB values
    Returns
    -------
    str
        The hex color code of the input RGB color.
    """

    return "#" + "".join(
        ["0{:x}".format(v) if v < 16 else "{:x}".format(v) for v in rgb]
    )


def get_color_palette(img_url, tolerance, limit):
    """
    Gets the color palette of an image

    Extracts the most common colors from an image
    and returns them as a dataframe of hex color codes and RGB values

    Parameters
    ----------
    img_url: str
        the url of the image from which the color palette is to be extracted
    tolerance: int
        a value between 0 and 100 representing the tolerance level for color matching
    limit: int
        the maximum number of colors to be returned in the palette

    Returns
    ----------
    dataframe
        a dataframe of hex color codes and RGB values corresponding to the most common colors in the image

    Examples
    --------
    get_color_palette('https://i.imgur.com/s9egWBB.jpg', 20, 5)
    """

    if not check_param_validity(img_url, tolerance, limit):
        print("The input parameters are not valid")
        return

    temp_image_name = "image_for_extraction.jpg"
    temp_image_storage_path = r"./images"
    full_file_path = os.path.join(temp_image_storage_path, temp_image_name)
    resize_large_img = 1000

    # Send a GET request to the URL and get the image data
    try:
        response = requests.get(img_url, timeout=30)

    except requests.exceptions.ConnectionError:
        raise Exception(
            "There seems to be an error accessing the URL and downloading it as an image. Please check the URL and try again."
        )

    if not response.status_code == 200:
        raise Exception(response.reason)
        
    try:
        image = Image.open(BytesIO(response.content))

        if response.status_code != 200:
            raise Exception("Unable to get image from the web url.")

    except PIL.UnidentifiedImageError:
        raise Exception("The URL may not point to an image. Please check the URL and try again.")

    image.thumbnail((resize_large_img, resize_large_img))
    image.name = temp_image_name
    image.seek(0)

    # Make necessary directories to the required path
    if not os.path.exists(temp_image_storage_path):
        os.makedirs(temp_image_storage_path)

    image.save(full_file_path)

    colors, _ = extract_from_path(full_file_path, tolerance=tolerance, limit=limit)

    df = pd.DataFrame(columns=["HEX", "RGB"])
    data = {
        "HEX": [rgb_to_hex(rgb_value) for rgb_value in [color[0] for color in colors]],
        "RGB": [
            ",".join(map(str, rgb_value))
            for rgb_value in [color[0] for color in colors]
        ],
        "Color Count": [freq for freq in [color[1] for color in colors]],
    }
    df = pd.DataFrame(data)

    # Clean Up
    os.remove(full_file_path)

    return df


def donut(img_url, num_clrs, tolerance, img_size, plot_show=True):
    """Create a donut chart of the top n colors in the image (number of colors specified by the user)

    Creates a square donut chart using the n most common colors from the image

    Parameters
    ----------
    img_url: str
        the url of the image that the user is pulling the colors from
    num_clrs: int
        the number of colors the user wants to pull from the image
    tolerance: int
        a value between 0 and 100 representing the tolerance level for color matching
    img_size: int
        the pixel width and height of the resulting chart
    plot_show: bool
        set False to supress output of the chart

    Returns
    ----------
    matplotlib.figure.Figure
        Donut chart of the colours

    Examples
    --------
    donut('https://i.imgur.com/s9egWBB.jpg', 5, 20, 400)
    """

    # get the top 100 colors and their proportion in the image
    df = get_color_palette(img_url, tolerance, limit=100)
    colors_prop = [
        round(x / sum(df["Color Count"].to_list()), 2)
        for x in df["Color Count"].to_list()
    ]

    img_colors = df["HEX"].to_list()[0:num_clrs]
    img_colors.append("#a9a9a9")

    value = colors_prop[0:num_clrs]
    value.append(round(1 - sum(value), 2))

    # need to do this to keep the donut hole in order
    factor = 0  # initialize the factor
    img_size = img_size / 227  # macbook air resolution is 227 pixels/inch
    if img_size > 600 / 227:
        factor = 0.3
    elif img_size <= 200 / 227:
        factor = 0.7
    elif img_size <= 400 / 227:
        factor = 0.6
    elif img_size <= 600 / 227:
        factor = 0.4

    category_value = []
    for i in range(len(img_colors)):
        category_value.append(img_colors[i] + ": " + str(f"{value[i]*100:.0f}%"))

    # create the donut hole
    pie = plt.Circle((0, 0), radius=img_size * factor, color="white")

    # make labels
    labels = category_value[:-1]
    labels.append("Other colors: " + str(f"{value[-1]*100:.0f}%"))

    # label and color
    plt.pie(value, labels=labels, colors=img_colors, radius=img_size)
    p = plt.gcf()
    p.gca().add_artist(pie)

    if plot_show:
        plt.show()
        return
    return p


def scatterplot(img_url, dataset, x, y, fill, tolerance=50):
    """Create a two-dimensional scatterplot based on the colours of the image

    Creates a simple scatterplot using the colours selected from the image,
    plotting two features from a dataset of the users choosing.

    Parameters
    ----------
    img_url : str
        url of the image to extract colours
    dataset : pandas.DataFrame
        the dataset to plot (already imported)
    x: str
        the data to plot on the x-axis
    y: str
        the data to plot on the y-axis
    fill: str
        the data to use to fill in the points of the scatter plot
    tolerance : int
        number between 0 and 100 used to give better visual representation;
        0 will not group any similar colours together, 100 will group all colours into one

    Returns
    ----------
    altair.vegalite.v4.api.Chart
        Scatterplot using image colours

    Examples
    --------
    scatterplot('https://i.imgur.com/s9egWBB.jpg', penguins, 'bill_length_mm', 'body_mass_g', 'species', 50)
    """
    if not img_url.startswith("https://"):
        raise ValueError("'img_url' must be a link (not a path).")

    if not [ext for ext in [".png", ".jpg", ".jpeg"] if (ext in img_url)]:
        raise ValueError("'img_url' must be a direct link to an image file.")

    if not isinstance(dataset, pd.core.frame.DataFrame):
        raise TypeError("'dataset' must be a pandas DataFrame.")

    if not isinstance(x, str):
        raise TypeError("'x' must be a string value.")

    if not isinstance(y, str):
        raise TypeError("'y' must be a string value.")

    if not isinstance(fill, str):
        raise TypeError("'y' must be a string value.")

    if not 0 <= tolerance <= 100:
        raise ValueError("'tolerance' must be between 0 and 100.")

    # Get image colour palette
    colours = get_color_palette(img_url, tolerance, dataset[fill].nunique())

    # Make Scatterplot based of the image colour palette
    scatter = (
        alt.Chart(dataset)
        .mark_point()
        .encode(
            alt.X(shorthand=x),
            alt.Y(shorthand=y),
            alt.Fill(shorthand=fill, scale=alt.Scale(range=colours["HEX"].to_list())),
        )
    )
    return scatter


def negative(img_url, num_colours=1, tolerance=0):
    """Invert top n colours in an image file.

    Colours are extracted from an image via URL and reversed,
    (e.g. red becomes cyan, green becomes magenta, yellow becomes blue)
    then stored in a table as HEX codes and RGB values.

    Parameters
    ----------
    img_url : str
        URL of an image file
    num_colours : int
        number of colours to be extracted
    tolerance : int
        number between 0 and 100 used to give better visual representation;
        0 will not group any similar colours together, 100 will group all colours into one

    Returns
    -------
    pandas.DataFrame
        a table of the top n inverted colours in the image, including their HEX codes and RGB values

    Examples
    --------
    >>> negative("https://i.imgur.com/s9egWBB.jpg", 3, 20)
    """
    if not img_url.startswith("https://"):
        raise ValueError("'img_url' must be a link (not a path).")

    if not [ext for ext in [".png", ".jpg", ".jpeg"] if (ext in img_url)]:
        raise ValueError("'img_url' must be a direct link to an image file.")

    if not isinstance(num_colours, int):
        raise TypeError("'num_colours' must be an integer value.")

    if not isinstance(tolerance, int):
        raise TypeError("'tolerance' must be an integer value.")

    if not 0 <= tolerance <= 100:
        raise ValueError("'tolerance' must be between 0 and 100.")

    # Load image
    try:
        img = Image.open(BytesIO(requests.get(img_url, stream=True).content))
   
    except PIL.UnidentifiedImageError:
        raise Exception("URL has not been read. Try another URL.")

    # Resize image for processing - 800 pixel maximum
    width = 800 / float(img.size[0])
    height = int((float(img.size[1]) * float(width)))
    img = img.resize((800, height), Image.LANCZOS)

    # Extract colours
    colours, pixel_count = extcolors.extract_from_image(
        img, tolerance=tolerance, limit=num_colours
    )

    # Check if there are any non-transparent pixels
    if not colours:
        raise ValueError(
            "No coloured pixels detected in the image. It is likely transparent or something went wrong."
        )

    # Format RGB codes into list
    colour_list = str(colours).replace("[(", "").split(", (")
    rgb_list = [i.split("), ")[0] + ")" for i in colour_list]

    # Inverse RGB colour codes and extract HEX codes
    inversed_rgb = []
    hex = []
    for cols in rgb_list:
        rgb = cols.replace("(", "").replace(")", "").split(", ")
        inverse_code = (255 - int(rgb[0]), 255 - int(rgb[1]), 255 - int(rgb[2]))
        inversed_rgb.append(inverse_code)
        hex.append("#%02x%02x%02x" % inverse_code)

    # Format data frame
    df = pd.DataFrame({"HEX": hex, "RGB": inversed_rgb})

    return df
